Fix 536 3.9 urlparse changes

g-k opened the pull request:
Fixes #536 

Currently:

* bleach just uses `urlparse` to extract a URI scheme/protocol (via `urlsplit`)
* when `http` is in `allowed_schemes` (the default), bleach [treats unparseable or empty schemes as implicitly `http` and safe](https://github.com/mozilla/bleach/blob/f5971aa57461cce00d6dd474e901bcbd00f240f1/bleach/sanitizer.py#L493-L496).
  * AFAIK this aligns with the web [url parsing spec](https://url.spec.whatwg.org/#url-parsing) and [tests](https://github.com/web-platform-tests/wpt/blob/master/url/) treating relative URIs as having the document's `base`, `origin`, and `scheme`.

* CPython <3.9 `urlparse` treats all digits trailing the first colon in a URL as a port number and puts it in the parsed URL path
* CPython 3.9 `urlparse` treats everything up to the first colon in the URI as the scheme https://github.com/python/cpython/pull/16837/files#diff-b3712475a413ec972134c0260c8f1eb1deefb66184f740ef00c37b4487ef873eL448-R433

  * bleach relied on the special port number parsing for URIs without a scheme `localhost:8000` and worked around it to handle custom schemes/protocols: https://github.com/mozilla/bleach/blob/f5971aa57461cce00d6dd474e901bcbd00f240f1/bleach/sanitizer.py#L489-L491

Considered a few options:

* ignore it: mark the failing tests as expected failures on 3.9. This gets tests passing, but introduces a breaking change for 3.9 users and doesn't fix the underlying issue.
* vendor an old urlparse version: either vendor [2.7.18](https://github.com/python/cpython/blob/v2.7.18/Lib/urlparse.py) and make it compatible with PY3 or drop Python 3 support and vendor a urlparse version <3.9. This wouldn't introduce breaking changes, but we're vendoring more code
* make the URI sanitizer stricter: stop doing that. This is stricter and more secure, but a breaking change that deviants from the web spec and probably how users expect it to work
* enumerate bad schemes: check for schemes that can deliver malicious content (e.g. `javascript:`, `file:`, `data:`) and block them. Requires adding a blocklist, keeping it up to date, and knowing all potentially evil schemes, which [isn't recommended](https://www.ranum.com/security/computer_security/editorials/dumb/)  
* match schemes that look like network locations: check if schemes look like valid hostnames or IPv4 or v6 addresses with an optional port (using bits of [django's URL validator](https://github.com/django/django/blob/354c1524b38c9b9f052c1d78dcbfa6ed5559aeb3/django/core/validators.py#L65)). Potentially breaking change and might confuse users if they expect `allowed_schemes` to behave the same as their stdlib `urlparse`.

This PR implements the last option using the hostname, IP address, and port portions of Django's URL validator and updates CI and tox to test against 3.9.

Open questions:

Is it safe to trust the Django URL validator for HTML contexts? I think so, or at least I'm not aware of any net locations that are also XSS vectors.

Is the Django code compatible?  It's [BSD 3 licensed in setup.cfg](https://github.com/django/django/blob/master/setup.cfg#L9), which should be compatible with MPL-2 (but also I'm not a lawyer).

If it is compatible, how should we attribute it in `bleach`?

willkg commented:
Why is the option "vendor all of urlparse" and not "implement a minimal `get_schema` that does what Bleach needs it to"? Is the latter option difficult for some reason I'm not seeing?

willkg commented:
Oops--scratch what I said. I missed stuff in the code changes.

willkg commented on bleach/utils.py:
```diff
@@ -40,3 +41,72 @@ def force_unicode(text):
     return six.text_type(text, "utf-8", "strict")
+
+
+# a subset of Django's URL validator
```
If this code is being copied from Django, it should be in a standalone module with the copyright, license information, and some information about which version was copied at the top. Maybe it's better maintenance-wise to put it in `vendor/`.

g-k commented:
OK!

* moved the django bits to the `_vendor` namespace 
* updated the vendoring scripts so we don't ship all of django
* switched to vendoring django 1.11 since it support 2.7+ (confirmed it uses the same regexes)

r? @willkg

willkg commented:
@g-k Can you rebase this? Then I'll test it out.

g-k commented:
Yeah I need to bump the version again.

g-k commented:
OK rebased

willkg commented on bleach/utils.py:
```diff
@@ -40,3 +43,52 @@ def force_unicode(text):
+    # one byte for the length of the name and one byte for the trailing dot
+    # that's used to indicate absolute names in DNS.
+    netloc = netloc_port_re.match(netloc)
+    return bool(netloc and len(netloc.group(0)) < 254)
```
While this is mostly based on the Django code, I think this is worth adding tests for.

g-k replied:
Dropped the django direction.

willkg commented on bleach/utils.py:
```diff
@@ -40,3 +43,52 @@ def force_unicode(text):
+                break
+        return scheme
+
+    return None
```
I think I'd add tests for this, too.

willkg replied:
Something like:

```
@pytest.mark.parametrize("text, expected", [
    ("", None),
    ("noscheme", None),
    ("https://example.com", "https"),
    ("git+ssh://example.com", "git+ssh"),
    # FIXME: example for . and -
])
def test_parse_uri_scheme(data, expected):
    assert _parse_uri_scheme(data) == expected
```

g-k replied:
Ended up not trying to pull in the urlparse 3.9 changes.

willkg commented on setup.py:
```diff
@@ -61,6 +61,7 @@ def get_version():
         'Programming Language :: Python :: 3.6',
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: 3.8',
+        'Programming Language :: Python :: 3.9',
```
Hooray!

willkg commented on tests/test_clean.py:
```diff
@@ -494,6 +511,11 @@ def test_attributes_list():
             '<a href="#example.com">foo</a>',
```
This test is confusing (probably my fault) since it uses a domain-ish thing, but it's testing anchors.

g-k replied:
Agreed! Added another less confusing looking anchor test and comment on the domain-ish test.

willkg reviewed (commented):
This looks good! I had some minor issues, but I like the approach.

g-k commented:
FWIW Freddy pointed out that Django forked urlsplit https://github.com/django/django/blob/main/django/utils/http.py#L287

g-k commented on tests/test_clean.py:
```diff
@@ -522,6 +544,14 @@ def test_attributes_list():
             {"protocols": ["http"]},
             '<a href="192.168.100.100:8000">valid</a>',
         ),
+        pytest.param(
```
I'll file an issue and see if anyone runs into this in the wild.

g-k commented:
Finally got a chance to spend some time on this.

This PR now:

* vendors the python 3.6.14 `urllib.parse`, uses it on all python versions (3.9 is the only one using the updated behavior, but it's simpler to not have version specific behavior); so it drops custom URL parsing behavior and django regexes
* updates vendor install and verify scripts (including the changes from https://github.com/mozilla/bleach/pull/565#discussion_r568149377 and https://github.com/mozilla/bleach/pull/565#discussion_r568156034; I don't think the code the other comments were on)

We probably should switch to newer URL parsing code eventually, but this PR keeps the changes minimal to fix the Python 3.9 test failures and avoid breaking changes.

Functional tests (not covered by CI):
- [ ] run `vendor_install.sh` and make sure nothing changes
- [ ] run `vendor_verify.sh` and make sure it passes
- [ ] change `parse.py` then run `vendor_verify.sh` and make sure it fails

willkg commented:
@g-k I need some time to re-load Bleach things into my noggin so as to really go through this. I think I can get to this next week.

g-k commented:
> @g-k I need some time to re-load Bleach things into my noggin so as to really go through this. I think I can get to this next week.

Thanks! I'll be away and won't get back to this until the end of the month anyway, so there's no rush.
 
I changed directions a few times on this, so the history is more involved and convoluted than it needs to be.

willkg commented on requirements-dev.txt:
```diff
@@ -17,3 +17,9 @@ wheel
+black; implementation_name == "cpython" and python_version > "3.5"
+
+# Requirements for type-check most recently mypy==0.790 and mypy-extensions==0.4.3
+mypy; implementation_name == "cpython" and python_version > "3.5"
```
Bleach doesn't support Python < 3.6, so can we nix the python version check here?

g-k replied:
Yep! Dropped it in an updated https://github.com/mozilla/bleach/pull/565/commits/84451282132955ffa6941b3cb0ff77a31ff0d749

willkg reviewed (approved):
What a messy situation!

I think this is the right thing to do. It's straight-forward, easy to reason about, and similar to what other projects decided to do. If things come up later, this probably won't make changing our minds later harder. Nicely done!

g-k commented on bleach/_vendor/pip_install_vendor.sh:
```diff
@@ -1,4 +1,25 @@
+DEST=${DEST:-"$(pwd)"}
+echo "installing vendored packages into ${DEST}"
+
+pip install --upgrade --no-binary all --no-compile --no-deps -r "${SCRIPTPATH}/vendor.txt" --target "${DEST}"
```
Dropped for the updated PR since it wasn't required and potentially introduced breakage.

g-k commented on bleach/_vendor/pip_install_vendor.sh:
```diff
@@ -1,4 +1,25 @@
+echo "cleaning files in ${DEST}/django"
+find "${DEST}/django" -type f -not -wholename "${DEST}/django/core/validators.py" -delete
+find "${DEST}/django" -empty -type d -delete
+touch "${DEST}/django/__init__.py" "${DEST}/django/core/__init__.py"
```
Ended up not going the Django route.

g-k commented on bleach/_vendor/vendor.txt:
```diff
@@ -1,3 +1,6 @@
+Django==1.11.29 \
+    --hash=sha256:014e3392058d94f40569206a24523ce254d55ad2f9f46c6550b0fe2e4f94cf3f \
+    --hash=sha256:4200aefb6678019a0acf0005cd14cfce3a5e6b9b90d06145fcdd2e474ad4329c
```
Ended up not going the Django route.

g-k commented on bleach/utils.py:
```diff
@@ -40,3 +43,52 @@ def force_unicode(text):
+
+# Characters valid in scheme names
+scheme_chars = (
+    "abcdefghijklmnopqrstuvwxyz" "ABCDEFGHIJKLMNOPQRSTUVWXYZ" "0123456789" "+-."
```
N/A. Ended up vendoring the whole urlparse library instead.

g-k commented on bleach/utils.py:
```diff
@@ -40,3 +43,52 @@ def force_unicode(text):
+
+def _is_valid_netloc_and_port(netloc):
+    """
+    Returns the scheme for a URI or None when parsing the URI fails
```
Dropped this change.

g-k commented on CHANGES:
```diff
@@ -1,6 +1,18 @@
 Bleach changes
 ==============
 
+Version 4.1.0 (August 25th, 2021)
```
Minor version bump since we're adding a new vendored dep.

g-k commented:
Thanks @willkg! I appreciate your feedback on all the various iterations of the PR.
